apollo-router/src/apollo_studio_interop/mod.rs

```diff
@@ -19,12 +21,131 @@ use apollo_compiler::executable::InlineFragment;
+#[derive(Clone, Default, Debug, Serialize)]
+pub(crate) struct InputObjectFieldStats {
+    /// True if the input object field was referenced.
+    pub(crate) referenced: bool,
+    /// True if the input object field was referenced but the value was null.
+    pub(crate) null_reference: bool,
+    /// True if the input object field was missing or undefined.
+    pub(crate) undefined_reference: bool,
+}
+
+/// The stats for a an input object field across multiple executions.
+#[derive(Clone, Default, Debug, Serialize)]
+pub(crate) struct AggregatedInputObjectFieldStats {
+    /// The number of executions where the field was referenced.
+    pub(crate) referenced: u64,
+    /// The number of executions where the field was referenced with a null value.
+    pub(crate) null_reference: u64,
+    /// The number of executions where the field was missing or undefined.
+    pub(crate) undefined_reference: u64,
+}
+
+/// The result of the generate_extended_references function which contains input object field and
+/// enum value stats for a single execution.
+#[derive(Clone, Default, Debug, Serialize)]
+pub(crate) struct ExtendedReferenceStats {
+    /// A map of parent type to a map of field name to stats
+    pub(crate) referenced_input_fields: HashMap<String, HashMap<String, InputObjectFieldStats>>,
+    /// A map of enum name to a set of enum values that were referenced
+    pub(crate) referenced_enums: HashMap<String, HashSet<String>>,
+}
+
+impl ExtendedReferenceStats {
+    pub(crate) fn new() -> Self {
+        ExtendedReferenceStats {
+            referenced_input_fields: HashMap::new(),
+            referenced_enums: HashMap::new(),
+        }
+    }
+}
+
+/// The aggregation of ExtendedReferenceStats across a number of executions.
+#[derive(Clone, Default, Debug, Serialize)]
+pub(crate) struct AggregatedExtendedReferenceStats {
+    /// A map of parent type to a map of field name to aggregated stats
+    pub(crate) referenced_input_fields:
+        HashMap<String, HashMap<String, AggregatedInputObjectFieldStats>>,
+    /// A map of enum name to a map of enum values to the number of executions that referenced them.
+    pub(crate) referenced_enums: HashMap<String, HashMap<String, u64>>,
+}
+
+impl AggregatedExtendedReferenceStats {
+    pub(crate) fn is_empty(&self) -> bool {
+        self.referenced_input_fields.is_empty() && self.referenced_enums.is_empty()
+    }
+}
+
+impl AddAssign<ExtendedReferenceStats> for AggregatedExtendedReferenceStats {
+    fn add_assign(&mut self, other_stats: ExtendedReferenceStats) {
+        // Merge input object references
+        for (type_name, type_stats) in other_stats.referenced_input_fields.iter() {
+            for (field_name, field_stats) in type_stats.iter() {
+                match self
+                    .referenced_input_fields
+                    .entry(type_name.to_string())
+                    .or_default()
+                    .entry(field_name.to_string())
```

Comment:
if we're in a hot path here. We already spotted that using the entry api might be less performant